Description: Improve GNU/Linux Makefile
 Make sure to use all the build flags that can be expected in the
 environment, overriding the maintainer build flags if necessary;
 split the build flags among the variables in the standard way;
 correctly declare all dependencies between source files; install
 localized man pages in the expected locations; get rid of the
 symlink hackery.
Forwarded: Marco Peereboom <marco@peereboom.us>
Author: Andrea Bolognani <eof@kiyuko.org>
Last-Update: 2012-02-19
Index: spectrwm.git/linux/Makefile
===================================================================
--- spectrwm.git.orig/linux/Makefile	2012-02-19 21:45:56.712437363 +0100
+++ spectrwm.git/linux/Makefile	2012-02-19 21:48:32.302582574 +0100
@@ -1,53 +1,69 @@
-CFLAGS+= -Wall -ggdb -D_GNU_SOURCE -I.
-CFLAGS+= -DSWM_LIB=\"$(LIBDIR)/libswmhack.so.$(LVERS)\"
-LDADD+= -lX11 -lXrandr -lXtst
+PREFIX       ?= /usr/local
+BINDIR       ?= $(PREFIX)/bin
+LIBDIR       ?= $(PREFIX)/lib
+DATAROOTDIR  ?= $(PREFIX)/share
+MANDIR       ?= $(DATAROOTDIR)/man
+XSESSIONSDIR ?= $(DATAROOTDIR)/xsessions
+
+BUILDVERSION = $(shell sh $(CURDIR)/../buildver.sh)
+LIBVERSION   = $(shell . $(CURDIR)/../lib/shlib_version; echo $$major.$$minor)
+
+MAINT_CFLAGS   = -Wall -ggdb
+MAINT_CPPFLAGS = -I. -D_GNU_SOURCE -DSWM_LIB=\"$(LIBDIR)/libswmhack.so.$(LIBVERSION)\"
+MAINT_LDLIBS   = -lX11 -lXrandr -lXtst
 
-PREFIX?= /usr/local
-BINDIR?= $(PREFIX)/bin
-LIBDIR?= $(PREFIX)/lib
-MANDIR?= $(PREFIX)/share/man
-
-CC= gcc
-
-LVERS= $(shell . ../lib/shlib_version; echo $$major.$$minor)
-
-BUILDVERSION= $(shell sh $(CURDIR)/../buildver.sh)
 ifneq ("${BUILDVERSION}", "")
-CFLAGS+= -DSPECTRWM_BUILDSTR=\"$(BUILDVERSION)\"
+MAINT_CPPFLAGS += -DSPECTRWM_BUILDSTR=\"$(BUILDVERSION)\"
 endif
 
-all: spectrwm libswmhack.so.$(LVERS)
+all: spectrwm libswmhack.so.$(LIBVERSION)
 
-spectrwm.c:
-	ln -sf ../spectrwm.c
-	ln -sf ../version.h
+spectrwm: spectrwm.o linux.o
+	$(CC) $(MAINT_LDFLAGS) $(LDFLAGS) -o $@ $+ $(MAINT_LDLIBS) $(LDLIBS)
 
-swm_hack.c:
-	ln -sf ../lib/swm_hack.c
+spectrwm.o: ../spectrwm.c ../version.h tree.h util.h
+	$(CC) $(MAINT_CFLAGS) $(CFLAGS) $(MAINT_CPPFLAGS) $(CPPFLAGS) -c $< -o $@
 
-spectrwm: spectrwm.o linux.o
-	$(CC) $(LDFLAGS) -o $@ $+ $(LDADD)
+linux.o: linux.c util.h
+	$(CC) $(MAINT_CFLAGS) $(CFLAGS) $(MAINT_CPPFLAGS) $(CPPFLAGS) -c $< -o $@
 
-%.so: %.c
-	$(CC) $(CFLAGS) -c -fpic -DPIC $+ -o $@
+libswmhack.so.$(LIBVERSION): swm_hack.so
+	$(CC) $(MAINT_LDFLAGS) $(LDFLAGS) -shared -fpic $+ -o $@ $(MAINT_LDLIBS) $(LDLIBS)
 
-libswmhack.so.$(LVERS): swm_hack.so
-	$(CC) -shared -fpic -o libswmhack.so.$(LVERS) swm_hack.so $(LDADD)
+swm_hack.so: ../lib/swm_hack.c
+	$(CC) $(MAINT_CFLAGS) $(CFLAGS) $(MAINT_CPPFLAGS) $(CPPFLAGS) -c -fpic -DPIC $< -o $@
+
+clean:
+	rm -f spectrwm *.o libswmhack.so.* *.so
 
 install: all
 	install -m 755 -d $(DESTDIR)$(BINDIR)
 	install -m 755 -d $(DESTDIR)$(LIBDIR)
 	install -m 755 -d $(DESTDIR)$(MANDIR)/man1
-	install -m 755 spectrwm $(DESTDIR)$(BINDIR)
-	install -m 755 libswmhack.so.$(LVERS) $(DESTDIR)$(LIBDIR)
-	install -m 644 ../spectrwm.1 $(DESTDIR)$(MANDIR)/man1/spectrwm.1
-	install -m 644 ../spectrwm_es.1 $(DESTDIR)$(MANDIR)/man1/spectrwm_es.1
-	install -m 644 ../spectrwm_it.1 $(DESTDIR)$(MANDIR)/man1/spectrwm_it.1
-	install -m 644 ../spectrwm_pt.1 $(DESTDIR)$(MANDIR)/man1/spectrwm_pt.1
-	install -m 644 ../spectrwm_ru.1 $(DESTDIR)$(MANDIR)/man1/spectrwm_ru.1
-	ln -sf $(DESTDIR)$(BINDIR)/spectrwm $(DESTDIR)$(BINDIR)/scrotwm
-
-clean:
-	rm -f spectrwm *.o *.so libswmhack.so.* spectrwm.c swm_hack.c version.h
+	install -m 755 -d $(DESTDIR)$(MANDIR)/es/man1
+	install -m 755 -d $(DESTDIR)$(MANDIR)/it/man1
+	install -m 755 -d $(DESTDIR)$(MANDIR)/pt/man1
+	install -m 755 -d $(DESTDIR)$(MANDIR)/ru/man1
+	install -m 755 -d $(DESTDIR)$(XSESSIONSDIR)
+	install -m 755 spectrwm                    $(DESTDIR)$(BINDIR)
+	ln -sf ../spectrwm                         $(DESTDIR)$(BINDIR)/scrotwm
+	install -m 644 libswmhack.so.$(LIBVERSION) $(DESTDIR)$(LIBDIR)
+	install -m 644 ../spectrwm.1               $(DESTDIR)$(MANDIR)/man1/spectrwm.1
+	install -m 644 ../spectrwm_es.1            $(DESTDIR)$(MANDIR)/es/man1/spectrwm.1
+	install -m 644 ../spectrwm_it.1            $(DESTDIR)$(MANDIR)/it/man1/spectrwm.1
+	install -m 644 ../spectrwm_pt.1            $(DESTDIR)$(MANDIR)/pt/man1/spectrwm.1
+	install -m 644 ../spectrwm_ru.1            $(DESTDIR)$(MANDIR)/ru/man1/spectrwm.1
+	install -m 644 spectrwm.desktop            $(DESTDIR)$(XSESSIONSDIR)
+
+uninstall:
+	rm -f $(DESTDIR)$(BINDIR)/spectrwm
+	rm -f $(DESTDIR)$(BINDIR)/scrotwm
+	rm -f $(DESTDIR)$(LIBDIR)/libswmhack.so.$(LIBVERSION)
+	rm -f $(DESTDIR)$(MANDIR)/man1/spectrwm.1
+	rm -f $(DESTDIR)$(MANDIR)/es/man1/spectrwm.1
+	rm -f $(DESTDIR)$(MANDIR)/it/man1/spectrwm.1
+	rm -f $(DESTDIR)$(MANDIR)/pt/man1/spectrwm.1
+	rm -f $(DESTDIR)$(MANDIR)/ru/man1/spectrwm.1
+	rm -f $(DESTDIR)$(XSESSIONSDIR)/spectrwm.desktop
 
-.PHONY: all install clean
+.PHONY: all clean install uninstall
